File: Examples/ObjC/Sources/Location/Readme.m
#import "ReadmeAction.h"

@import CoreLocation;

#pragma mark - Internal

static CLLocationManager *_locationManager;

CLLocationManager *locationManager(void) {
  if (!_locationManager) {
    _locationManager = [[CLLocationManager alloc] init];
  }
  return _locationManager;
}

#if TARGET_OS_IOS

void requestAlwaysAuthorization(ReadmeViewController *viewController) {
  [locationManager() requestAlwaysAuthorization];
}

#endif

void requestWhenInUseAuthorization(ReadmeViewController *viewController) {
  [locationManager() requestWhenInUseAuthorization];
}

#pragma mark - Readme

NSString *const readme =
    @"This sample presents a complete BrazeLocation integration.\n"
    @"\n"
    @"- AppDelegate.{h,m}:\n"
    @"  - Configure the braze instance with BrazeLocationProvider";

#if TARGET_OS_IOS

NSInteger const actionsCount = 2;

ReadmeAction const actions[] = {
    {@"Request \"always\" authorization", @"",
     ^(ReadmeViewController *_Nonnull viewController){
         requestAlwaysAuthorization(viewController);
}
}
, {@"Request \"when in use\" authorization", @"",
   ^(ReadmeViewController *_Nonnull viewController){
       requestWhenInUseAuthorization(viewController);
}
}
,
}
;

#elif TARGET_OS_TV

NSInteger const actionsCount = 1;

ReadmeAction const actions[] = {
    {@"Request \"when in use\" authorization", @"",
     ^(ReadmeViewController *_Nonnull viewController){
         requestWhenInUseAuthorization(viewController);
}
}
,
}
;

#endif

File: Examples/Swift/Sources/Location/Readme.swift
import CoreLocation

let readme =
  """
  This sample presents a complete BrazeLocation integration which enables location tracking and geofence monitoring.

  - AppDelegate.swift:
    - Configure the braze instance with BrazeLocationProvider
  """

#if os(iOS)
  @MainActor
  let actions: [(String, String, @MainActor (ReadmeViewController) -> Void)] = [
    (#"Request "always" authorization"#, "", requestAlwaysAuthorization),
    (#"Request "when in use" authorization"#, "", requestWhenInUseAuthorization),
  ]
#elseif os(tvOS) || os(visionOS)
  @MainActor
  let actions: [(String, String, @MainActor (ReadmeViewController) -> Void)] = [
    (#"Request "when in use" authorization"#, "", requestWhenInUseAuthorization)
  ]
#endif

// MARK: - Internal

@MainActor
let locationManager = CLLocationManager()

#if os(iOS)
  @MainActor
  func requestAlwaysAuthorization(_ viewController: ReadmeViewController) {
    locationManager.requestAlwaysAuthorization()
  }
#endif

@MainActor
func requestWhenInUseAuthorization(_ viewController: ReadmeViewController) {
  locationManager.requestWhenInUseAuthorization()
}
